raft/src/main/scala/zio/raft/PendingCommands.scala

```diff
@@ -1,26 +1,23 @@
 package zio.raft
 
-import zio.{Ref, ZIO}
+import zio.{UIO, ZIO}
 
-case class PendingCommands(ref: Ref[Map[Index, Any]]):
-  def complete[R](index: Index, response: R) =
-    for
-      maybePromise <- ref
-        .modify(map => (map.get(index), map.removed(index)))
-        .map(_.asInstanceOf[Option[CommandPromise[R]]])
-      _ <- ZIO.foreachDiscard(maybePromise)(promise => promise.succeed(response))
-    yield ()
+case class PendingCommands(map: Map[Index, Any]):
+  def withCompleted[R](index: Index, response: R): UIO[PendingCommands] =
+    ZIO
+      .foreach(map.get(index).asInstanceOf[Option[CommandPromise[R]]])(_.succeed(response))
+      .as(PendingCommands(map.removed(index)))
 
-  def add[R](index: Index, promise: CommandPromise[R]) =
-    ref.update(_ + (index -> promise))
+  def withAdded[R](index: Index, promise: CommandPromise[R]): PendingCommands =
+    PendingCommands(map + (index -> promise))
 
-  def reset(leaderId: Option[MemberId]) =
-    for
-      promises <- ref.getAndSet(Map.empty).map(_.values.map(_.asInstanceOf[CommandPromise[Any]]))
-      _ <- ZIO.foreach(promises)(_.fail(NotALeaderError(leaderId)))
-    yield ()
+  def stepDown(leaderId: Option[MemberId]): UIO[Unit] =
+    ZIO
+      .foreach(map.values.map(_.asInstanceOf[CommandPromise[Any]]))(_.fail(NotALeaderError(leaderId)))
+      .unit
+
+  // TODO (eran): optimize this by using the same structure as pendingReads and taking last
+  def lastIndex: Option[Index] = Option.when(map.keys.nonEmpty)(map.keys.max(using Ordering.by(_.value)))
```

Comment:
can we have a unit test for this? that it is actually taking the maximum number?